cpp/src/slice2java/Gen.h

```diff
@@ -8,89 +8,129 @@
 
 namespace Slice
 {
+    //
+    // Generate code to marshal or unmarshal a type.
+    //
+    enum OptionalMode
+    {
+        OptionalNone,
+        OptionalInParam,
+        OptionalOutParam,
+        OptionalReturnParam,
+        OptionalMember
+    };
+
     class JavaVisitor : public JavaGenerator, public ParserVisitor
     {
     public:
         ~JavaVisitor() override;
 
-    protected:
-        JavaVisitor(const std::string&);
-
-        enum ParamDir
-        {
-            InParam,
-            OutParam
-        };
+        static void writeMarshalUnmarshalCode(
+            IceInternal::Output& out,
+            const std::string& package,
+            const TypePtr& type,
+            OptionalMode mode,
+            bool optionalMapping,
+            std::int32_t tag,
+            const std::string& param,
+            bool marshal,
+            int& iter,
+            const std::string& customStream = "",
+            const MetadataList& metadata = MetadataList(),
+            const std::string& patchParams = "");
 
-        std::string getResultType(const OperationPtr&, const std::string&, bool, bool);
```

Comment:
Otherwise, the rest of these functions are preserved exactly as they were, except I added `static` to the ones which didn't access any member variables. That's the only change though. I didn't even fix the comments, or add parameter names in this PR. Baby steps.